File: computing/interview/oop/cpp/shape_example.cpp
#include <iostream>
#include <memory>
#include <vector>
#include <cmath>

class Shape {
public:
    virtual ~Shape() = default;
    virtual double area() const = 0;
    virtual void describe() const = 0;
};

class Circle : public Shape {
public:
    explicit Circle(double radius) : radius_(radius) {}

    double area() const override {
        return M_PI * radius_ * radius_;
    }

    void describe() const override {
        std::cout << "Circle radius=" << radius_ << " area=" << area() << "\n";
    }

private:
    double radius_;
};

class Rectangle : public Shape {
public:
    Rectangle(double width, double height) : width_(width), height_(height) {}

    double area() const override {
        return width_ * height_;
    }

    void describe() const override {
        std::cout << "Rectangle " << width_ << "x" << height_ << " area=" << area() << "\n";
    }

private:
    double width_;
    double height_;
};

int main() {
    std::vector<std::shared_ptr<Shape>> shapes;
    shapes.emplace_back(std::make_shared<Circle>(2.0));
    shapes.emplace_back(std::make_shared<Rectangle>(3.0, 4.0));

    for (const auto& shape : shapes) {
        shape->describe();
    }

    return 0;
}

File: computing/interview/oop/python/shape_example.py
from __future__ import annotations
from abc import ABC, abstractmethod
from math import pi
from typing import List


class Shape(ABC):
    @abstractmethod
    def area(self) -> float:
        ...

    @abstractmethod
    def describe(self) -> None:
        ...


class Circle(Shape):
    def __init__(self, radius: float) -> None:
        self._radius = radius

    def area(self) -> float:
        return pi * self._radius * self._radius

    def describe(self) -> None:
        print(f"Circle radius={self._radius} area={self.area():.3f}")


class Rectangle(Shape):
    def __init__(self, width: float, height: float) -> None:
        self._width = width
        self._height = height

    def area(self) -> float:
        return self._width * self._height

    def describe(self) -> None:
        print(f"Rectangle {self._width}x{self._height} area={self.area():.3f}")


def main() -> None:
    shapes: List[Shape] = [Circle(2.0), Rectangle(3.0, 4.0)]
    for shape in shapes:
        shape.describe()


if __name__ == "__main__":
    main()
